hw/ip/otbn/dv/otbnsim/sim/wsr.py

```diff
@@ -0,0 +1,131 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import List, Optional
+
+from riscvmodel.types import Trace  # type: ignore
+
+
+class TraceWSR(Trace):  # type: ignore
+    def __init__(self, wsr_name: str, new_value: int):
+        self.wsr_name = wsr_name
+        self.new_value = new_value
+
+    def __str__(self) -> str:
+        return '{} = {:#x}'.format(self.wsr_name, self.new_value)
+
+
+class WSR:
+    '''Models a Wide Status Register'''
+    def __init__(self, name: str):
+        self.name = name
+
+    def read_unsigned(self) -> int:
+        '''Get the stored value as a 256-bit unsigned value'''
+        raise NotImplementedError()
+
+    def write_unsigned(self, value: int) -> None:
+        '''Set the stored value as a 256-bit unsigned value'''
+        raise NotImplementedError()
+
+    def read_signed(self) -> int:
+        '''Get the stored value as a 256-bit signed value'''
+        uval = self.read_unsigned()
+        return uval - (1 << 256 if uval >> 255 else 0)
+
+    def write_signed(self, value: int) -> None:
+        '''Set the stored value as a 256-bit signed value'''
+        assert -(1 << 255) <= value < (1 << 255)
+        uval = (1 << 256) + value if value < 0 else value
+        self.write_unsigned(uval)
+
+    def commit(self) -> None:
+        '''Commit pending changes'''
+        return
+
+    def changes(self) -> List[TraceWSR]:
+        '''Return list of pending architectural changes'''
+        return []
+
+
+class DumbWSR(WSR):
+    '''Models a WSR without special behaviour'''
+    def __init__(self, name: str):
+        super().__init__(name)
+        self._value = 0
+        self._next_value = None  # type: Optional[int]
+
+    def read_unsigned(self) -> int:
+        return self._value
+
+    def write_unsigned(self, value: int) -> None:
+        assert 0 <= value < (1 << 256)
+        self._next_value = value
+
+    def commit(self) -> None:
+        if self._next_value is not None:
+            self._value = self._next_value
+        self._next_value = None
+
+    def changes(self) -> List[TraceWSR]:
+        return ([TraceWSR(self.name, self._next_value)]
+                if self._next_value is not None
+                else [])
+
+
+class RandWSR(WSR):
+    '''The magic RND WSR'''
+    def __init__(self, name: str, seed: int):
+        super().__init__(name)
+        # RND has its own generator, independent of the rest of the process, to
+        # make it easier to ensure things stay reproducible.
+        #
+        # TODO: We've got to work out how to make the RTL and ISS do the same
+        #       thing. Probably model the RTL's LFSR here, and allow seeding to
+        #       simulate the CSRNG/EDN? Again, it's going to be more helpful to
+        #       have a local generator either way.
+        self.generator = random.Random(seed)
```

Comment:
In the RTL I just use a fixed value for now, perhaps just replicate that fixed value here?

Whilst it's trivial in the simulator to produce some 'real' random numbers, it isn't in the RTL and at this stage of the design simply testing reading the RND WSR and checking it gets you a value is sufficient testing and having the ISS and RTL match behaviour here seems useful.

Reply:
Sounds like a good idea. Can you point me at where the magic random value is set?

Reply:
It's set in the OTBN top level at ip/otbn/rtl/otbn_core.sv:

```
assign rnd = 256'h257a986a406da3799e49a7e32b7668f3b4e4a36a5ad34847c0b568978318a930;
```

Though that's only present in the wide ALU PR at the moment: https://github.com/lowRISC/opentitan/pull/3471

Add a comment as well that says it must be kept in sync with the RTL and I'll do the same in the RTL in that PR.

Reply:
Felix' simulator sets the RND to 1. Whatever we do, it would be great at the moment to pick a static value at the moment. The algorithms use randomness in various places (including to influence control flow), and the current bringup e.g. of RSA is much easier if we can compare exactly what the different implementations do on an instruction-by-instruction level.

Reply:
Oh and I should also note the CSR read takes the bottom 32-bits of that value: `0x8318a930`